### PR TITLE
Support reordering RVar if a function is associative.

### LINKED ISSUE #3214: Support reordering RVar if the function is associative
Halide currently doesn't allow `.reorder` schedule on RVar hitting the following assertion. It will be useful to allow reordering reduction loops if the the function is proven to be associative.

https://github.com/halide/Halide/blob/8c2f090f40411cd370ae0dd80b98f88367a7ad38/src/Func.cpp#L1510-L1524

I think we could reuse some existing tool implemented for `.rfactor` to prove the associativity.

### PR DESCRIPTION
**Pull request by jingpu**
Fixes #3214.

### CODE REVIEW

File: test/correctness/reorder_rvars.cpp [hunk 1]
```diff
@@ -24,7 +24,9 @@ int main(int argc, char **argv) {
 
         // Reorder g
         g.reorder(y, x);
-        g.update(0).reorder(r1.x, y, x, r1.y);
+        // It is legal to reorder r1.x and r1.y
+        // because stage g.update(0) is associative.
+        g.update(0).reorder(r1.y, y, x, r1.x);
```
Comment (steven-johnson) [inline test/correctness/reorder_rvars.cpp]: Might be worth adding a negative test in test/errors for the is-not-associative case

Comment (jingpu) [reply test/correctness/reorder_rvars.cpp]: There is an "error_bad_rvar_order" test that still works with this PR. I think it gives an okay coverage. Let me know if you want me to add another test.

File: src/Func.cpp [hunk 2]
```diff
@@ -1536,7 +1541,15 @@ void reorder_vars(vector<Dim> &dims_old, const VarOrRVar *vars, size_t size, con
 }
 
 Stage &Stage::reorder(const std::vector<VarOrRVar>& vars) {
-    reorder_vars(definition.schedule().dims(), &vars[0], vars.size(), *this);
+    const string &func_name = function.name();
+    vector<Expr> &args = definition.args();
+    vector<Expr> &values = definition.values();
+
+    // Check whether the operator is associative
+    const auto &prover_result = prove_associativity(func_name, args, values);
```
Comment (abadams) [inline src/Func.cpp]: Reordering also changes the order in which the reductions happen, as well as reassociating them, so I think you additionally need commutative.

**Comment by abadams** [comment]
@psuriana Does this look correct to you?

**Comment by jingpu** [comment]
I added the check for commutative.

**Comment by abadams** [comment]
Cool. It looks correct to me, but I'd like to hear from Patricia in case I'm missing something.

**Comment by psuriana** [comment]
LGTM

File: test/correctness/reorder_rvars.cpp (hunk 1 above)
Comment (steven-johnson) [reply test/correctness/reorder_rvars.cpp]: Nope, that's good enough

**Comment by steven-johnson** [comment]
Failure is unrelated, merging

**Pull request merged**